packages/ra-core/src/routing/useGetPathForRecordCallback.ts

```diff
@@ -23,46 +26,74 @@ export const useGetPathForRecordCallback = <
                 );
             }
             const resourceDefinition = resourceDefinitions[finalResource] ?? {};
-            const linkFunc = typeof link === 'function' ? link : () => link;
-
-            const defaultLink = resourceDefinition.hasShow
-                ? 'show'
-                : resourceDefinition.hasEdit
-                  ? 'edit'
-                  : false;
-
-            const isLinkFalse =
-                link === false || (link == null && defaultLink === false);
 
-            if (record == null || isLinkFalse) {
+            if (record == null || link === false) {
                 return false;
             }
-            const linkResult = linkFunc(record, finalResource) ?? defaultLink;
-            const linkResultIsPromise = isPromise(linkResult);
 
-            if (linkResultIsPromise) {
-                return linkResult.then(resolvedLink => {
-                    if (resolvedLink === false) {
-                        // already set to false by default
-                        return;
-                    }
+            // When the link prop is not provided, we infer a default value and check whether users
+            // can access it
+            if (link == null) {
+                if (
+                    resourceDefinition.hasShow &&
+                    (await canAccess({
+                        action: 'show',
+                        resource: finalResource,
+                        record,
+                    }))
+                ) {
+                    return createPath({
+                        resource: finalResource,
+                        id: record.id,
+                        type: 'show',
+                    });
+                }
+                if (
+                    resourceDefinition.hasEdit &&
+                    (await canAccess({
+                        action: 'edit',
+                        resource: finalResource,
+                        record,
+                    }))
+                ) {
                     return createPath({
                         resource: finalResource,
                         id: record.id,
-                        type: resolvedLink,
+                        type: 'edit',
                     });
```

Comment:
This is nice, but this code will do two subsequent calls to `canAccess` in many cases. Can we do the two checks (`edit` and `show`) in parallel, then set the default value?